File: alienblob.js
function AlienBlobDrawer(width, height, numColors) {
  this.name = "AlienBlob";
  this.pos = Math.floor(Math.random() * 1e6);
  this.colorIndex = 0;
  this.values = {speed: 20, colorSpeed: 10, detail: 3, zoom: 70};
  this.ranges = {speed: [0,100], colorSpeed: [0,100], detail: [1,6], zoom: [0,100]};
  this.decay = 0.5;
  this.speedMultiplier = 0.07;
  
  this.alienblob = new AlienBlob(width, height, numColors);
}

AlienBlobDrawer.prototype.draw = function(leds, palette) {
  var indices = this.alienblob.run(this.pos, this.values["zoom"]/100.0, this.values["detail"], this.decay);
  //console.log(indices);
  
  for (var x=0; x<leds.width; x++) {
    for (var y=0; y<leds.height; y++) {
      index = indices[x*leds.height + y] + this.colorIndex;
      leds.rgbs[x][y] = palette.rgbs[index % palette.numColors];
    }
  }
  
  this.pos += this.speedMultiplier * this.values.speed / 100.0;
  this.colorIndex++;
}

AlienBlobDrawer.prototype.getDelay = function() {
  return 1000/this.values.speed;
}



function AlienBlob(width, height, numColors) {
  this.width = width;
  this.height = height;
  this.numColors = numColors;
   
  // create precalculated sine table
  this.sineTable = [];
  for (i = 0; i < 360; i++)
    this.sineTable[i] = Math.sin(i * Math.PI / 180);
  
  this.indices = [];
  
  this.perlin = new PerlinNoise();
}

AlienBlob.prototype.run = function(zoff, zoom, detail, decay) {
  var incr = 0.3125;
  var noiseMult = 7;
  var multiplier = (1-zoom) + 0.02;

  var y, x;
  var yy = 0;
  for (var y = 0; y < this.height; y++) {
    var xx = 0;
    for (var x = 0; x < this.width; x++) {
      var n = this.perlin.noise(xx*multiplier, yy*multiplier, zoff, detail, decay);
      // use pre-calculated sine results
      var deg = Math.floor((n * noiseMult + 4 * Math.PI) * 180 / Math.PI);
      var h = (this.sineTable[deg % 360] + 1) / 2;
      
      // determine pixel color
      this.indices[x*this.height + y] = Math.floor(h * this.numColors);
      
      xx += incr;
    }
    
    yy += incr;
  }
  
  return this.indices;
}



/* -*- mode: java; c-basic-offset: 2; indent-tabs-mode: nil -*- */

/*
 Part of the Processing project - http://processing.org
 
 Copyright (c) 2012-13 The Processing Foundation
 Copyright (c) 2004-12 Ben Fry and Casey Reas
 Copyright (c) 2001-04 Massachusetts Institute of Technology
 
 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation, version 2.1.
 
 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 Lesser General Public License for more details.
 
 You should have received a copy of the GNU Lesser General
 Public License along with this library; if not, write to the
 Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 Boston, MA  02111-1307  USA
 */
 
 // Translated to javascript by Greg Friedland 9/2/2014

//static float noise_fsc(float i);

//#define RANDOM() (random() / (float)RAND_MAX)

SINCOS_PRECISION = 0.5;
SINCOS_LENGTH = (360*2);
DEG_TO_RAD = (3.14159 / 180);

PERLIN_YWRAPB = 4;
PERLIN_YWRAP = (1<<PERLIN_YWRAPB);
PERLIN_ZWRAPB = 8;
PERLIN_ZWRAP = (1<<PERLIN_ZWRAPB);
PERLIN_SIZE = 4095;

//static float sinLUT[SINCOS_LENGTH];
//static float cosLUT[SINCOS_LENGTH];
//static int perlin_TWOPI, perlin_PI;
//static float *perlin_cosTable;
//static float perlin[PERLIN_SIZE+1];


function PerlinNoise() {
  //srandom(clock());
  
  this.perlin_TWOPI = SINCOS_LENGTH;
  this.perlin_PI = this.perlin_TWOPI / 2;
  
  this.cosTable = [];
  for (i = 0; i < SINCOS_LENGTH; i++) {
    //sinLUT[i] = (float) sin(i * DEG_TO_RAD * SINCOS_PRECISION);
    this.cosTable[i] = Math.cos(i * DEG_TO_RAD * SINCOS_PRECISION);
  }

  this.perlin = [];
  for (i = 0; i < PERLIN_SIZE + 1; i++) {
    this.perlin[i] = Math.random();
  }
}

PerlinNoise.prototype.noise_fsc = function(i) {
  // using bagel's cosine table instead
  return 0.5*(1.0-this.cosTable[Math.floor(i*this.perlin_PI)%this.perlin_TWOPI]);
}

PerlinNoise.prototype.noise = function(x, y, z, perlin_octaves, perlin_amp_falloff) {
  if (x<0) x=-x;
  if (y<0) y=-y;
  if (z<0) z=-z;
  
  var xi=Math.floor(x), yi=Math.floor(y), zi=Math.floor(z);
  var xf = x - xi;
  var yf = y - yi;
  var zf = z - zi;
  var rxf, ryf;
  
  var r=0;
  var ampl=0.5;
  
  var n1,n2,n3;
  
  for (var i=0; i<perlin_octaves; i++) {
    var of=xi+(yi<<PERLIN_YWRAPB)+(zi<<PERLIN_ZWRAPB);
    
    rxf = this.noise_fsc(xf);
    ryf = this.noise_fsc(yf);
    
    n1  = this.perlin[of&PERLIN_SIZE];
    n1 += rxf*(this.perlin[(of+1)&PERLIN_SIZE]-n1);
    n2  = this.perlin[(of+PERLIN_YWRAP)&PERLIN_SIZE];
    n2 += rxf*(this.perlin[(of+PERLIN_YWRAP+1)&PERLIN_SIZE]-n2);
    n1 += ryf*(n2-n1);

    of += PERLIN_ZWRAP;
    n2  = this.perlin[of&PERLIN_SIZE];
    n2 += rxf*(this.perlin[(of+1)&PERLIN_SIZE]-n2);
    n3  = this.perlin[(of+PERLIN_YWRAP)&PERLIN_SIZE];
    n3 += rxf*(this.perlin[(of+PERLIN_YWRAP+1)&PERLIN_SIZE]-n3);
    n2 += ryf*(n3-n2);
    
    n1 += this.noise_fsc(zf)*(n2-n1);
    
    r += n1*ampl;
    ampl *= perlin_amp_falloff;
    xi<<=1; xf*=2;
    yi<<=1; yf*=2;
    zi<<=1; zf*=2;
    
    if (xf>=1.0) { xi++; xf--; }
    if (yf>=1.0) { yi++; yf--; }
    if (zf>=1.0) { zi++; zf--; }
  }
  return r;
}

module.exports.AlienBlobDrawer = AlienBlobDrawer;

File: alienblob.cc
#include <node.h>
#include <v8.h>
#include <nan.h>

#include <stdio.h>
#include <stdlib.h>

#include <math.h>
#include <stdint.h>
#include <time.h>

#include <dlfcn.h>
#include <unistd.h>

#define MAX(a,b) ((a) > (b) ? (a) : (b))
#define MIN(a,b) ((a) < (b) ? (a) : (b))
#define PI 3.1415926

static float sineTable[360];


static void init_noise(void);
static float noise(float x, float y, float z, int octaves, float decay);
// static float pnoise_harmonics(float x, float y, float z, int octaves, float decay);

static void alienblob(int width, int height, int numColors, float zoff, int perlinOctaves,
                      float perlinDecay, float zoom, int *indices);



class AlienBlob : public Nan::ObjectWrap {
 public:
    static void Init(v8::Local<v8::Object> exports) {
        Nan::HandleScope scope;

        // Prepare constructor template
        v8::Local<v8::FunctionTemplate> tpl = Nan::New<v8::FunctionTemplate>(New);
        tpl->SetClassName(Nan::New("AlienBlob").ToLocalChecked());
        tpl->InstanceTemplate()->SetInternalFieldCount(0);

        // Prototype
        Nan::SetPrototypeMethod(tpl, "run", Run);

        constructor.Reset(tpl->GetFunction());
        exports->Set(Nan::New("AlienBlob").ToLocalChecked(), tpl->GetFunction());
    }

 private:
    explicit AlienBlob() {}
    ~AlienBlob() {}

    static void New(const Nan::FunctionCallbackInfo<v8::Value>& info) {
        if (info.IsConstructCall()) {
            // Invoked as constructor: `new AlienBlob(...)`
            AlienBlob* obj = new AlienBlob();
            obj->Wrap(info.This());
            info.GetReturnValue().Set(info.This());
        } else {
            // Invoked as plain function `AlienBlob(...)`, turn into construct call.
            const int argc = 0;
            v8::Local<v8::Value> argv[argc] = { };
            v8::Local<v8::Function> cons = Nan::New<v8::Function>(constructor);
            info.GetReturnValue().Set(cons->NewInstance(argc, argv));
        }

        init_noise();
        for (int i = 0; i < 360; i ++) sineTable[i] = sin(i * PI / 180);
    }

    static void Run(const Nan::FunctionCallbackInfo<v8::Value>& info) {
        int width = info[0]->NumberValue();
        int height = info[1]->NumberValue();
        int numColors = info[2]->NumberValue();
        float zoff = info[3]->NumberValue();
        int perlinOctaves = info[4]->NumberValue();
        float perlinDecay = info[5]->NumberValue();
        float zoom = info[6]->NumberValue();

        v8::Local<v8::Object> array = info[7]->ToObject();
        v8::Handle<v8::Object> buffer = array->Get(v8::String::New("buffer"))->ToObject();
        unsigned int offset = array->Get(v8::String::New("byteOffset"))->Uint32Value();
        // int length = array->Get(v8::String::New("byteLength"))->Uint32Value();
        int* indices = (int*) &((char*) buffer->GetIndexedPropertiesExternalArrayData())[offset];

        alienblob(width, height, numColors, zoff, perlinOctaves, perlinDecay, zoom, indices);

        info.GetReturnValue().Set(Nan::Undefined());
    }

    static Nan::Persistent<v8::Function> constructor;
};

Nan::Persistent<v8::Function> AlienBlob::constructor;

void InitAll(v8::Local<v8::Object> exports) {
    AlienBlob::Init(exports);
}

NODE_MODULE(alienblob, InitAll)







static void alienblob(int width, int height, int numColors, float zoff, int perlinOctaves,
                             float perlinDecay, float zoom, int *indices) {
    float incr = 0.3125;
    float noiseMult = 7;
    float multiplier = (1-zoom) + 0.02;

    int y, x;
    float yy = 0;
    for (y = 0; y < height; y++) {
        float xx = 0;
        for (x = 0; x < width; x++) {
            //float n = pnoise_harmonics(xx*multiplier, yy*multiplier, zoff,
            //                           perlinOctaves, perlinDecay);
            float n = noise(xx*multiplier, yy*multiplier, zoff, perlinOctaves, perlinDecay);
            // use pre-calculated sine results
            int deg = (int) ((n * noiseMult + 4 * PI) * 180 / PI);
            float h = (sineTable[deg % 360] + 1) / 2;
            
            // determine pixel color
            indices[x*height + y] = (int) (h * numColors);
            
            xx += incr;
        }
        
        yy += incr;
    }  
}




#if 0
/*
 Ken Perlins improved noise   -  http://mrl.nyu.edu/~perlin/noise/
 C-port:  http://www.fundza.com/c4serious/noise/perlin/perlin.html
 by Malcolm Kesson;   arduino port by Peter Chiochetti, Sep 2007 :
 -  make permutation constant byte, obsoletes init(), lookup % 256
 */


static const uint8_t p[] = {   151,160,137,91,90, 15,131, 13,201,95,96,
    53,194,233, 7,225,140,36,103,30,69,142, 8,99,37,240,21,10,23,190, 6,
    148,247,120,234,75, 0,26,197,62,94,252,219,203,117, 35,11,32,57,177,
    33,88,237,149,56,87,174,20,125,136,171,168,68,175,74,165,71,134,139,
    48,27,166, 77,146,158,231,83,111,229,122, 60,211,133,230,220,105,92,
    41,55,46,245,40,244,102,143,54,65,25,63,161, 1,216,80,73,209,76,132,
    187,208, 89, 18,169,200,196,135,130,116,188,159, 86,164,100,109,198,
    173,186, 3,64,52,217,226,250,124,123,5,202,38,147,118,126,255,82,85,
    212,207,206, 59,227, 47,16,58,17,182,189, 28,42,223,183,170,213,119,
    248,152,2,44,154,163,70,221,153,101,155,167,43,172, 9,129,22,39,253,
    19,98,108,110,79,113,224,232,178,185,112,104,218,246, 97,228,251,34,
    242,193,238,210,144,12,191,179,162,241,81,51,145,235,249,14,239,107,
    49,192,214, 31,181,199,106,157,184, 84,204,176,115,121,50,45,127, 4,
    150,254,138,236,205, 93,222,114, 67,29,24, 72,243,141,128,195,78,66,
    215,61,156,180
};

static float fade(float t){ return t * t * t * (t * (t * 6 - 15) + 10); }
static float lerp(float t, float a, float b){ return a + t * (b - a); }
static float grad(int hash, float x, float y, float z)
{
    int     h = hash & 15;          /* CONVERT LO 4 BITS OF HASH CODE */
    float  u = h < 8 ? x : y,      /* INTO 12 GRADIENT DIRECTIONS.   */
    v = h < 4 ? y : h==12||h==14 ? x : z;
    return ((h&1) == 0 ? u : -u) + ((h&2) == 0 ? v : -v);
}

#define P(x) p[(x) & 255]

static float pnoise(float x, float y, float z)
{
    int   X = (int)floor(x) & 255,             /* FIND UNIT CUBE THAT */
    Y = (int)floor(y) & 255,             /* CONTAINS POINT.     */
    Z = (int)floor(z) & 255;
    x -= floor(x);                             /* FIND RELATIVE X,Y,Z */
    y -= floor(y);                             /* OF POINT IN CUBE.   */
    z -= floor(z);
    float  u = fade(x),                       /* COMPUTE FADE CURVES */
    v = fade(y),                       /* FOR EACH OF X,Y,Z.  */
    w = fade(z);
    
    //if (X==0 && Y==0 && x==0 && y==0) {
    //  Serial.print(X); Serial.print(" "); Serial.print(x); Serial.print(" "); Serial.print(Y); Serial.print(" ");
    //  Serial.print(y); Serial.print(" "); Serial.print(Z); Serial.print(" "); Serial.print(z); Serial.print(": "); Serial.println(w);
    //}
    
    int  A = P(X)+Y,
    AA = P(A)+Z,
    AB = P(A+1)+Z,                        /* HASH COORDINATES OF */
    B = P(X+1)+Y,
    BA = P(B)+Z,
    BB = P(B+1)+Z;                        /* THE 8 CUBE CORNERS, */
    
    return lerp(w,lerp(v,lerp(u, grad(P(AA  ), x, y, z),   /* AND ADD */
                                                        grad(P(BA  ), x-1, y, z)),   /* BLENDED */
                                         lerp(u, grad(P(AB  ), x, y-1, z),        /* RESULTS */
                                                    grad(P(BB  ), x-1, y-1, z))),       /* FROM  8 */
                            lerp(v, lerp(u, grad(P(AA+1), x, y, z-1),  /* CORNERS */
                                                     grad(P(BA+1), x-1, y, z-1)),          /* OF CUBE */
                                     lerp(u, grad(P(AB+1), x, y-1, z-1),
                                                grad(P(BB+1), x-1, y-1, z-1))));
}

// x, y, z, num harmonics, unused, harmonic falloff
static float pnoise_harmonics(float x, float y, float z, int octaves, float decay) {
    float scale = 0;
    float result = 0;
    float decay2 = 1;
    int i;
    for (i=0; i<octaves; i++ ) {
        scale += decay2;
        result += pnoise(x, y, z) * decay2;
        decay2 *= decay;
        //    x *= freq;
        //    y *= freq;
    }
    
    return result / scale;
}
#endif





/* -*- mode: java; c-basic-offset: 2; indent-tabs-mode: nil -*- */

/*
 Part of the Processing project - http://processing.org
 
 Copyright (c) 2012-13 The Processing Foundation
 Copyright (c) 2004-12 Ben Fry and Casey Reas
 Copyright (c) 2001-04 Massachusetts Institute of Technology
 
 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation, version 2.1.
 
 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 Lesser General Public License for more details.
 
 You should have received a copy of the GNU Lesser General
 Public License along with this library; if not, write to the
 Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 Boston, MA  02111-1307  USA
 */

static float noise_fsc(float i);

#define RANDOM() (random() / (float)RAND_MAX)

#define SINCOS_PRECISION 0.5f
#define SINCOS_LENGTH (360*2)
#define DEG_TO_RAD (3.14159 / 180)

#define PERLIN_YWRAPB 4
#define PERLIN_YWRAP (1<<PERLIN_YWRAPB)
#define PERLIN_ZWRAPB 8
#define PERLIN_ZWRAP (1<<PERLIN_ZWRAPB)
#define PERLIN_SIZE 4095

static float sinLUT[SINCOS_LENGTH];
static float cosLUT[SINCOS_LENGTH];
static int perlin_TWOPI, perlin_PI;
static float *perlin_cosTable;
static float perlin[PERLIN_SIZE+1];


static void init_noise(void) {
    srandom(clock());
    
    perlin_TWOPI = perlin_PI = SINCOS_LENGTH;
    perlin_PI >>= 1;
    
    int i;
    for (i = 0; i < SINCOS_LENGTH; i++) {
        sinLUT[i] = (float) sin(i * DEG_TO_RAD * SINCOS_PRECISION);
        cosLUT[i] = (float) cos(i * DEG_TO_RAD * SINCOS_PRECISION);
    }
    perlin_cosTable = cosLUT;
    
    for (i = 0; i < PERLIN_SIZE + 1; i++) {
        perlin[i] = RANDOM(); //perlinRandom.nextFloat(); //(float)Math.random();
    }
}

inline float noise_fsc(float i) {
    // using bagel's cosine table instead
    return 0.5f*(1.0f-perlin_cosTable[(int)(i*perlin_PI)%perlin_TWOPI]);
}

static float noise(float x, float y, float z, int perlin_octaves, float perlin_amp_falloff) {
    if (x<0) x=-x;
    if (y<0) y=-y;
    if (z<0) z=-z;
    
    int xi=(int)x, yi=(int)y, zi=(int)z;
    float xf = x - xi;
    float yf = y - yi;
    float zf = z - zi;
    float rxf, ryf;
    
    float r=0;
    float ampl=0.5f;
    
    float n1,n2,n3;
    
    int i;
    for (i=0; i<perlin_octaves; i++) {
        int of=xi+(yi<<PERLIN_YWRAPB)+(zi<<PERLIN_ZWRAPB);
        
        rxf=noise_fsc(xf);
        ryf=noise_fsc(yf);
        
        n1  = perlin[of&PERLIN_SIZE];
        n1 += rxf*(perlin[(of+1)&PERLIN_SIZE]-n1);
        n2  = perlin[(of+PERLIN_YWRAP)&PERLIN_SIZE];
        n2 += rxf*(perlin[(of+PERLIN_YWRAP+1)&PERLIN_SIZE]-n2);
        n1 += ryf*(n2-n1);

        of += PERLIN_ZWRAP;
        n2  = perlin[of&PERLIN_SIZE];
        n2 += rxf*(perlin[(of+1)&PERLIN_SIZE]-n2);
        n3  = perlin[(of+PERLIN_YWRAP)&PERLIN_SIZE];
        n3 += rxf*(perlin[(of+PERLIN_YWRAP+1)&PERLIN_SIZE]-n3);
        n2 += ryf*(n3-n2);
        
        n1 += noise_fsc(zf)*(n2-n1);
        
        r += n1*ampl;
        ampl *= perlin_amp_falloff;
        xi<<=1; xf*=2;
        yi<<=1; yf*=2;
        zi<<=1; zf*=2;
        
        if (xf>=1.0f) { xi++; xf--; }
        if (yf>=1.0f) { yi++; yf--; }
        if (zf>=1.0f) { zi++; zf--; }
    }
    return r;
}

int main(void) {
    init_noise();

    int i;
    for (i=0; i<544*30*10 * 10; i++) {
        noise(5.5, 4.1, 0.2, 4, 0.5);
        printf(".");
    }
    return 0;
}
